dstore.go

```
@@ -60,35 +60,62 @@ func (b *batch) GetTransaction() (*sql.Tx, error) {
 	return newTransaction, nil
 }
 
-func (b *batch) Put(key ds.Key, val []byte) error {
+func (b *batch) Put(key ds.Key, val []byte) (err error) {
+	defer func() {
+		if err != nil {
```

I'm not 100% sure what the behavior of `recover` is, but we might be able to clean this up by starting this function with:

```
if b.txn == nil {
    return
}
```

---

Good catch. You should be able to. The main thing I wanted to avoid was swallowing panics without re-panicking. If we return before calling `recover`, we won't be swallowing anything.